app/models/media/object.rb

```diff
@@ -45,19 +45,58 @@ class ::Media::Object < ApplicationRecord
 
     scope :expired, -> { where(answer_id: nil).where("created_at < ?", 12.hours.ago) }
 
-    after_save :generate_media_object_filename, if: :saved_change_to_answer_id?
+    after_save :generate_media_object_filename
 
     def dynamic_thumb?
       false
     end
 
-    private
-
     # Set a useful filename to assist data analysts who deal with lots of downloads.
     def generate_media_object_filename
+      return if item.record.answer_id.nil?
+      answer = item.record.answer
+      filename = "nemo-#{answer.response.shortcode}"
+      filename = build_filename(filename, item)
+      item.blob.update!(filename: filename)
+    end
+
+    # build a more complex filename if is nested in repeat groups
+    def build_filename(filename, item)
+      repeat_groups = []
       answer = item.record.answer
-      extension = File.extname(item.filename.to_s)
-      item.blob.update!(filename: "elmo-#{answer.response.shortcode}-#{answer.id}#{extension}")
+      answer_group = nil
+      answer_group = next_agroup_up(answer.parent_id) if answer.from_group?
+      if answer_group.present? && answer_group.repeatable?
+        repeat_groups = respect_ancestors(answer_group, repeat_groups)
+        filename += "-#{repeat_groups.pop}" until repeat_groups.empty?
+      else
+        filename += "_#{answer.question.code}"
+      end
+      filename += File.extname(item.filename.to_s)
+      filename
+    end
+
+    # returns an array of group name strings from all nested groups
+    def respect_ancestors(answer_group, repeat_groups)
+      name = answer_group.group_name.gsub(/\s+/, "_").to_s
```

Comment:
Quick suggestion: Rather than replacing characters here, replace in the final output of `build_filename` above, and catch more things (not just whitespace but probably special characters too so they're legal filenames).

What do you think about accepting `[0-9a-zA-Z_-.]`?